In [1]:
import pandas as pd

In [2]:
#Importing credit history csv
df = pd.read_csv("credit_record.csv")

In [3]:
df.head()

ID  MONTHS_BALANCE STATUS
0  5001711               0      X
1  5001711              -1      0
2  5001711              -2      0
3  5001711              -3      0
4  5001712               0      C

In [4]:
df.shape

(1048575, 3)

In [5]:
#Importing credit card applicant csv
app = pd.read_csv("application_record.csv")

In [6]:
app.head()

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0  5008804           M            Y               Y             0   
1  5008805           M            Y               Y             0   
2  5008806           M            Y               Y             0   
3  5008808           F            N               Y             0   
4  5008809           F            N               Y             0   

   AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0          427500.0               Working               Higher education   
1          427500.0               Working               Higher education   
2          112500.0               Working  Secondary / secondary special   
3          270000.0  Commercial associate  Secondary / secondary special   
4          270000.0  Commercial associate  Secondary / secondary special   

     NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0        Civil marriage   Rented apartment      -12005          -4542   
1        Civil marriage   Rented apartment      -12005          -4542   
2               Married  House / apartment      -21474          -1134   
3  Single / not married  House / apartment      -19110          -3051   
4  Single / not married  House / apartment      -19110          -3051   

   FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0           1                1           0           0             NaN   
1           1                1           0           0             NaN   
2           1                0           0           0  Security staff   
3           1                0           1           1     Sales staff   
4           1                0           1           1     Sales staff   

   CNT_FAM_MEMBERS  
0              2.0  
1              2.0  
2              2.0  
3              1.0  
4              1.0

In [7]:
app.shape

(438557, 18)

In [8]:
app.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 438557 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   ID                   438557 non-null  int64  
 1   CODE_GENDER          438557 non-null  object 
 2   FLAG_OWN_CAR         438557 non-null  object 
 3   FLAG_OWN_REALTY      438557 non-null  object 
 4   CNT_CHILDREN         438557 non-null  int64  
 5   AMT_INCOME_TOTAL     438557 non-null  float64
 6   NAME_INCOME_TYPE     438557 non-null  object 
 7   NAME_EDUCATION_TYPE  438557 non-null  object 
 8   NAME_FAMILY_STATUS   438557 non-null  object 
 9   NAME_HOUSING_TYPE    438557 non-null  object 
 10  DAYS_BIRTH           438557 non-null  int64  
 11  DAYS_EMPLOYED        438557 non-null  int64  
 12  FLAG_MOBIL           438557 non-null  int64  
 13  FLAG_WORK_PHONE      438557 non-null  int64  
 14  FLAG_PHONE           438557 non-null  int64  
 15  FLAG_EMAIL       

In [9]:
app['OCCUPATION_TYPE'].value_counts()

Laborers                 78240
Core staff               43007
Sales staff              41098
Managers                 35487
Drivers                  26090
High skill tech staff    17289
Accountants              15985
Medicine staff           13520
Cooking staff             8076
Security staff            7993
Cleaning staff            5845
Private service staff     3456
Low-skill Laborers        2140
Secretaries               2044
Waiters/barmen staff      1665
Realty agents             1041
HR staff                   774
IT staff                   604
Name: OCCUPATION_TYPE, dtype: int64

In [10]:
app[app['OCCUPATION_TYPE'].isnull()]

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0       5008804           M            Y               Y             0   
1       5008805           M            Y               Y             0   
7       5008812           F            N               Y             0   
8       5008813           F            N               Y             0   
9       5008814           F            N               Y             0   
...         ...         ...          ...             ...           ...   
438549  6840098           F            N               Y             0   
438550  6840100           F            N               Y             0   
438551  6840102           F            N               Y             0   
438552  6840104           M            N               Y             0   
438555  6842765           F            N               Y             0   

        AMT_INCOME_TOTAL NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0               427500.0          Working               Higher education   
1               427500.0          Working               Higher education   
7               283500.0        Pensioner               Higher education   
8               283500.0        Pensioner               Higher education   
9               283500.0        Pensioner               Higher education   
...                  ...              ...                            ...   
438549          135000.0        Pensioner  Secondary / secondary special   
438550          135000.0        Pensioner  Secondary / secondary special   
438551          135000.0        Pensioner  Secondary / secondary special   
438552          135000.0        Pensioner  Secondary / secondary special   
438555           72000.0        Pensioner  Secondary / secondary special   

       NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0          Civil marriage   Rented apartment      -12005          -4542   
1          Civil marriage   Rented apartment      -12005          -4542   
7               Separated  House / apartment      -22464         365243   
8               Separated  House / apartment      -22464         365243   
9               Separated  House / apartment      -22464         365243   
...                   ...                ...         ...            ...   
438549          Separated  House / apartment      -22717         365243   
438550          Separated  House / apartment      -22717         365243   
438551          Separated  House / apartment      -22717         365243   
438552          Separated  House / apartment      -22717         365243   
438555            Married  House / apartment      -21673         365243   

        FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0                1                1           0           0             NaN   
1                1                1           0           0             NaN   
7                1                0           0           0             NaN   
8                1                0           0           0             NaN   
9                1                0           0           0             NaN   
...            ...              ...         ...         ...             ...   
438549           1                0           0           0             NaN   
438550           1                0           0           0             NaN   
438551           1                0           0           0             NaN   
438552           1                0           0           0             NaN   
438555           1                0           0           0             NaN   

        CNT_FAM_MEMBERS  
0                   2.0  
1                   2.0  
7                   1.0  
8                   1.0  
9                   1.0  
...                 ...  
438549              1.0  
438550              1.0  
438551              1.0  
438552              1.0  
438555              2.0  

[134203 rows x 18 columns]

In [11]:
app.rename(columns={'CODE_GENDER': 'gender', 'FLAG_OWN_CAR': 'car_owner', 'FLAG_OWN_REALTY': 'property_owner', 'CNT_CHILDREN': 'children_count', 'AMT_INCOME_TOTAL': 'annual_income', 'NAME_INCOME_TYPE': 'income_type', 'NAME_EDUCATION_TYPE': 'education', 'NAME_FAMILY_STATUS': 'family_status', 'NAME_HOUSING_TYPE': 'housing_type', 'DAYS_BIRTH': 'days_from_birth', 'DAYS_EMPLOYED': 'days_employed', 'FLAG_MOBIL': 'mobile', 'FLAG_WORK_PHONE': 'work_phone', 'FLAG_PHONE': 'phone', 'FLAG_EMAIL': 'email', 'OCCUPATION_TYPE': 'occupation', 'CNT_FAM_MEMBERS': 'family_member_count'}, inplace=True)

In [12]:
app.head(1)

ID gender car_owner property_owner  children_count  annual_income  \
0  5008804      M         Y              Y               0       427500.0   

  income_type         education NAME_FAMILY_STATUS NAME_HOUSING_TYPE  \
0     Working  Higher education     Civil marriage  Rented apartment   

   DAYS_BIRTH  DAYS_EMPLOYED  FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  \
0      -12005          -4542           1                1           0   

   FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  
0           0             NaN              2.0

In [13]:
app.rename(columns={'NAME_FAMILY_STATUS': 'family_status', 'NAME_HOUSING_TYPE': 'housing_type', 'DAYS_BIRTH': 'days_from_birth', 'DAYS_EMPLOYED': 'days_employed', 'FLAG_MOBIL': 'mobile', 'FLAG_WORK_PHONE': 'work_phone', 'FLAG_PHONE': 'phone', 'FLAG_EMAIL': 'email', 'OCCUPATION_TYPE': 'occupation', 'CNT_FAM_MEMBERS': 'family_member_count'  }, inplace=True)

In [14]:
app.head(1)

ID gender car_owner property_owner  children_count  annual_income  \
0  5008804      M         Y              Y               0       427500.0   

  income_type         education   family_status      housing_type  \
0     Working  Higher education  Civil marriage  Rented apartment   

   days_from_birth  days_employed  mobile  work_phone  phone  email  \
0           -12005          -4542       1           1      0      0   

  occupation  family_member_count  
0        NaN                  2.0

In [15]:
app['days_from_birth'] = abs(app.days_from_birth)

In [16]:
# app['days_employed'] = abs(app.days_employed)

In [17]:
app.head(1)

ID gender car_owner property_owner  children_count  annual_income  \
0  5008804      M         Y              Y               0       427500.0   

  income_type         education   family_status      housing_type  \
0     Working  Higher education  Civil marriage  Rented apartment   

   days_from_birth  days_employed  mobile  work_phone  phone  email  \
0            12005          -4542       1           1      0      0   

  occupation  family_member_count  
0        NaN                  2.0

In [18]:
app['car_owner'] = app['car_owner'].map({'Y': 1, 'N': 0})


In [19]:
app['property_owner'] = app['property_owner'].map({'Y': 1, 'N': 0})


In [20]:
app.head(1)

ID gender  car_owner  property_owner  children_count  annual_income  \
0  5008804      M          1               1               0       427500.0   

  income_type         education   family_status      housing_type  \
0     Working  Higher education  Civil marriage  Rented apartment   

   days_from_birth  days_employed  mobile  work_phone  phone  email  \
0            12005          -4542       1           1      0      0   

  occupation  family_member_count  
0        NaN                  2.0

In [21]:
#Changing float columns to integers
cols = ['annual_income', 'family_member_count']
for col in cols:
   app[col] = app[col].apply(lambda x: int(x))

In [22]:
app.head(1)

ID gender  car_owner  property_owner  children_count  annual_income  \
0  5008804      M          1               1               0         427500   

  income_type         education   family_status      housing_type  \
0     Working  Higher education  Civil marriage  Rented apartment   

   days_from_birth  days_employed  mobile  work_phone  phone  email  \
0            12005          -4542       1           1      0      0   

  occupation  family_member_count  
0        NaN                    2

In [23]:
main = app.merge(df, on='ID', how='left')

In [24]:
main.head()

ID gender  car_owner  property_owner  children_count  annual_income  \
0  5008804      M          1               1               0         427500   
1  5008804      M          1               1               0         427500   
2  5008804      M          1               1               0         427500   
3  5008804      M          1               1               0         427500   
4  5008804      M          1               1               0         427500   

  income_type         education   family_status      housing_type  \
0     Working  Higher education  Civil marriage  Rented apartment   
1     Working  Higher education  Civil marriage  Rented apartment   
2     Working  Higher education  Civil marriage  Rented apartment   
3     Working  Higher education  Civil marriage  Rented apartment   
4     Working  Higher education  Civil marriage  Rented apartment   

   days_from_birth  days_employed  mobile  work_phone  phone  email  \
0            12005          -4542       1           1      0      0   
1            12005          -4542       1           1      0      0   
2            12005          -4542       1           1      0      0   
3            12005          -4542       1           1      0      0   
4            12005          -4542       1           1      0      0   

  occupation  family_member_count  MONTHS_BALANCE STATUS  
0        NaN                    2             0.0      C  
1        NaN                    2            -1.0      C  
2        NaN                    2            -2.0      C  
3        NaN                    2            -3.0      C  
4        NaN                    2            -4.0      C

In [25]:
main.rename(columns={'MONTHS_BALANCE': 'months_balance', 'STATUS': 'status'}, inplace=True)

In [26]:
main['months_balance'] = abs(main.months_balance)

In [27]:
main[(main['days_employed'] == 365243) & (main.income_type == 'Pensioner')]

ID gender  car_owner  property_owner  children_count  \
137      5008812      F          0               1               0   
138      5008812      F          0               1               0   
139      5008812      F          0               1               0   
140      5008812      F          0               1               0   
141      5008812      F          0               1               0   
...          ...    ...        ...             ...             ...   
1179807  6840098      F          0               1               0   
1179808  6840100      F          0               1               0   
1179809  6840102      F          0               1               0   
1179810  6840104      M          0               1               0   
1179813  6842765      F          0               1               0   

         annual_income income_type                      education  \
137             283500   Pensioner               Higher education   
138             283500   Pensioner               Higher education   
139             283500   Pensioner               Higher education   
140             283500   Pensioner               Higher education   
141             283500   Pensioner               Higher education   
...                ...         ...                            ...   
1179807         135000   Pensioner  Secondary / secondary special   
1179808         135000   Pensioner  Secondary / secondary special   
1179809         135000   Pensioner  Secondary / secondary special   
1179810         135000   Pensioner  Secondary / secondary special   
1179813          72000   Pensioner  Secondary / secondary special   

        family_status       housing_type  days_from_birth  days_employed  \
137         Separated  House / apartment            22464         365243   
138         Separated  House / apartment            22464         365243   
139         Separated  House / apartment            22464         365243   
140         Separated  House / apartment            22464         365243   
141         Separated  House / apartment            22464         365243   
...               ...                ...              ...            ...   
1179807     Separated  House / apartment            22717         365243   
1179808     Separated  House / apartment            22717         365243   
1179809     Separated  House / apartment            22717         365243   
1179810     Separated  House / apartment            22717         365243   
1179813       Married  House / apartment            21673         365243   

         mobile  work_phone  phone  email occupation  family_member_count  \
137           1           0      0      0        NaN                    1   
138           1           0      0      0        NaN                    1   
139           1           0      0      0        NaN                    1   
140           1           0      0      0        NaN                    1   
141           1           0      0      0        NaN                    1   
...         ...         ...    ...    ...        ...                  ...   
1179807       1           0      0      0        NaN                    1   
1179808       1           0      0      0        NaN                    1   
1179809       1           0      0      0        NaN                    1   
1179810       1           0      0      0        NaN                    1   
1179813       1           0      0      0        NaN                    2   

         months_balance status  
137                 4.0      0  
138                 5.0      X  
139                 6.0      X  
140                 7.0      0  
141                 8.0      0  
...                 ...    ...  
1179807             NaN    NaN  
1179808             NaN    NaN  
1179809             NaN    NaN  
1179810             NaN    NaN  
1179813             NaN    NaN  

[197166 rows x 20 columns]

In [28]:
##If they are receiving a pension, their days employed (actually days unemployed) == 365243

In [29]:
main[main.days_employed > 0]

ID gender  car_owner  property_owner  children_count  \
137      5008812      F          0               1               0   
138      5008812      F          0               1               0   
139      5008812      F          0               1               0   
140      5008812      F          0               1               0   
141      5008812      F          0               1               0   
...          ...    ...        ...             ...             ...   
1179807  6840098      F          0               1               0   
1179808  6840100      F          0               1               0   
1179809  6840102      F          0               1               0   
1179810  6840104      M          0               1               0   
1179813  6842765      F          0               1               0   

         annual_income income_type                      education  \
137             283500   Pensioner               Higher education   
138             283500   Pensioner               Higher education   
139             283500   Pensioner               Higher education   
140             283500   Pensioner               Higher education   
141             283500   Pensioner               Higher education   
...                ...         ...                            ...   
1179807         135000   Pensioner  Secondary / secondary special   
1179808         135000   Pensioner  Secondary / secondary special   
1179809         135000   Pensioner  Secondary / secondary special   
1179810         135000   Pensioner  Secondary / secondary special   
1179813          72000   Pensioner  Secondary / secondary special   

        family_status       housing_type  days_from_birth  days_employed  \
137         Separated  House / apartment            22464         365243   
138         Separated  House / apartment            22464         365243   
139         Separated  House / apartment            22464         365243   
140         Separated  House / apartment            22464         365243   
141         Separated  House / apartment            22464         365243   
...               ...                ...              ...            ...   
1179807     Separated  House / apartment            22717         365243   
1179808     Separated  House / apartment            22717         365243   
1179809     Separated  House / apartment            22717         365243   
1179810     Separated  House / apartment            22717         365243   
1179813       Married  House / apartment            21673         365243   

         mobile  work_phone  phone  email occupation  family_member_count  \
137           1           0      0      0        NaN                    1   
138           1           0      0      0        NaN                    1   
139           1           0      0      0        NaN                    1   
140           1           0      0      0        NaN                    1   
141           1           0      0      0        NaN                    1   
...         ...         ...    ...    ...        ...                  ...   
1179807       1           0      0      0        NaN                    1   
1179808       1           0      0      0        NaN                    1   
1179809       1           0      0      0        NaN                    1   
1179810       1           0      0      0        NaN                    1   
1179813       1           0      0      0        NaN                    2   

         months_balance status  
137                 4.0      0  
138                 5.0      X  
139                 6.0      X  
140                 7.0      0  
141                 8.0      0  
...                 ...    ...  
1179807             NaN    NaN  
1179808             NaN    NaN  
1179809             NaN    NaN  
1179810             NaN    NaN  
1179813             NaN    NaN  

[197166 rows x 20 columns]

In [30]:
main[main.days_employed == 0]

Empty DataFrame
Columns: [ID, gender, car_owner, property_owner, children_count, annual_income, income_type, education, family_status, housing_type, days_from_birth, days_employed, mobile, work_phone, phone, email, occupation, family_member_count, months_balance, status]
Index: []

In [31]:
#Apparently, no one applying for a credit card is unemployed. Only ones with negative days of employment are Pensioners. Also no one with zero days of employemnt

In [32]:
main[(main.days_employed > -13) & (main.days_employed < 0) ]

ID gender  car_owner  property_owner  children_count  \
1120121  6639411      F          0               1               1   
1120122  6639412      F          0               1               1   
1120123  6639413      F          0               1               1   
1120124  6639414      F          0               1               1   
1145978  6790284      F          0               0               0   
1145979  6790285      F          0               0               0   
1155523  7639412      F          0               1               1   
1166053  7790285      F          0               0               0   

         annual_income           income_type                      education  \
1120121          81000  Commercial associate  Secondary / secondary special   
1120122          81000  Commercial associate  Secondary / secondary special   
1120123          81000  Commercial associate  Secondary / secondary special   
1120124          81000  Commercial associate  Secondary / secondary special   
1145978         162000               Working               Higher education   
1145979         162000               Working               Higher education   
1155523          81000  Commercial associate  Secondary / secondary special   
1166053         162000               Working               Higher education   

        family_status       housing_type  days_from_birth  days_employed  \
1120121     Separated  House / apartment            11989            -12   
1120122     Separated  House / apartment            11989            -12   
1120123     Separated  House / apartment            11989            -12   
1120124     Separated  House / apartment            11989            -12   
1145978     Separated  House / apartment            17036            -12   
1145979     Separated  House / apartment            17036            -12   
1155523     Separated  House / apartment            11989            -12   
1166053     Separated  House / apartment            17036            -12   

         mobile  work_phone  phone  email  occupation  family_member_count  \
1120121       1           0      0      1    Laborers                    2   
1120122       1           0      0      1    Laborers                    2   
1120123       1           0      0      1    Laborers                    2   
1120124       1           0      0      1    Laborers                    2   
1145978       1           0      0      0  Core staff                    1   
1145979       1           0      0      0  Core staff                    1   
1155523       1           0      0      1    Laborers                    2   
1166053       1           0      0      0  Core staff                    1   

         months_balance status  
1120121             NaN    NaN  
1120122             NaN    NaN  
1120123             NaN    NaN  
1120124             NaN    NaN  
1145978             NaN    NaN  
1145979             NaN    NaN  
1155523             NaN    NaN  
1166053             NaN    NaN

In [33]:
#13 days is the minimum days of employment

In [34]:
main.status.value_counts()

C    329536
0    290654
X    145950
1      8747
5      1527
2       801
3       286
4       214
Name: status, dtype: int64

In [35]:
high_risk = main[main.status == '5']

In [36]:
len(high_risk)

1527

In [37]:
high_risk['ID'].value_counts().sort_values(ascending=False)

5142362    48
5142361    48
5085886    45
5117313    44
5116304    42
           ..
5054525     1
5052661     1
5050606     1
5050603     1
5023781     1
Name: ID, Length: 180, dtype: int64

In [38]:
high_risk.nunique()

ID                     180
gender                   2
car_owner                2
property_owner           2
children_count           5
annual_income           39
income_type              4
education                4
family_status            5
housing_type             6
days_from_birth        121
days_employed          102
mobile                   1
work_phone               2
phone                    2
email                    2
occupation              15
family_member_count      6
months_balance          54
status                   1
dtype: int64

In [39]:
#180 applicants with charged off accounts

In [40]:
len(high_risk[high_risk.ID == 5142362])

48

In [41]:
high_risk[high_risk.ID == 5142362]

ID gender  car_owner  property_owner  children_count  \
719923  5142362      M          1               1               1   
719924  5142362      M          1               1               1   
719925  5142362      M          1               1               1   
719926  5142362      M          1               1               1   
719927  5142362      M          1               1               1   
719928  5142362      M          1               1               1   
719929  5142362      M          1               1               1   
719930  5142362      M          1               1               1   
719931  5142362      M          1               1               1   
719932  5142362      M          1               1               1   
719933  5142362      M          1               1               1   
719934  5142362      M          1               1               1   
719935  5142362      M          1               1               1   
719936  5142362      M          1               1               1   
719937  5142362      M          1               1               1   
719938  5142362      M          1               1               1   
719939  5142362      M          1               1               1   
719940  5142362      M          1               1               1   
719941  5142362      M          1               1               1   
719942  5142362      M          1               1               1   
719943  5142362      M          1               1               1   
719944  5142362      M          1               1               1   
719945  5142362      M          1               1               1   
719946  5142362      M          1               1               1   
719947  5142362      M          1               1               1   
719948  5142362      M          1               1               1   
719949  5142362      M          1               1               1   
719950  5142362      M          1               1               1   
719951  5142362      M          1               1               1   
719952  5142362      M          1               1               1   
719953  5142362      M          1               1               1   
719954  5142362      M          1               1               1   
719955  5142362      M          1               1               1   
719956  5142362      M          1               1               1   
719957  5142362      M          1               1               1   
719958  5142362      M          1               1               1   
719959  5142362      M          1               1               1   
719960  5142362      M          1               1               1   
719961  5142362      M          1               1               1   
719962  5142362      M          1               1               1   
719963  5142362      M          1               1               1   
719964  5142362      M          1               1               1   
719965  5142362      M          1               1               1   
719966  5142362      M          1               1               1   
719967  5142362      M          1               1               1   
719968  5142362      M          1               1               1   
719969  5142362      M          1               1               1   
719970  5142362      M          1               1               1   

        annual_income income_type                      education  \
719923         112500     Working  Secondary / secondary special   
719924         112500     Working  Secondary / secondary special   
719925         112500     Working  Secondary / secondary special   
719926         112500     Working  Secondary / secondary special   
719927         112500     Working  Secondary / secondary special   
719928         112500     Working  Secondary / secondary special   
719929         112500     Working  Secondary / secondary special   
719930         112500     Working  Secondary / secondary special   
719931         1125

In [42]:
df[df.ID == 5142362]

ID  MONTHS_BALANCE STATUS
979818  5142362              -1      5
979819  5142362              -2      5
979820  5142362              -3      5
979821  5142362              -4      5
979822  5142362              -5      5
979823  5142362              -6      5
979824  5142362              -7      5
979825  5142362              -8      5
979826  5142362              -9      5
979827  5142362             -10      5
979828  5142362             -11      5
979829  5142362             -12      5
979830  5142362             -13      5
979831  5142362             -14      5
979832  5142362             -15      5
979833  5142362             -16      5
979834  5142362             -17      5
979835  5142362             -18      5
979836  5142362             -19      5
979837  5142362             -20      5
979838  5142362             -21      5
979839  5142362             -22      5
979840  5142362             -23      5
979841  5142362             -24      5
979842  5142362             -25      5
979843  5142362             -26      5
979844  5142362             -27      5
979845  5142362             -28      5
979846  5142362             -29      5
979847  5142362             -30      5
979848  5142362             -31      5
979849  5142362             -32      5
979850  5142362             -33      5
979851  5142362             -34      5
979852  5142362             -35      5
979853  5142362             -36      5
979854  5142362             -37      5
979855  5142362             -38      5
979856  5142362             -39      5
979857  5142362             -40      5
979858  5142362             -41      5
979859  5142362             -42      5
979860  5142362             -43      5
979861  5142362             -44      5
979862  5142362             -45      5
979863  5142362             -46      5
979864  5142362             -47      5
979865  5142362             -48      5
979866  5142362             -49      0
979867  5142362             -50      X
979868  5142362             -51      X
979869  5142362             -52      X
979870  5142362             -53      X
979871  5142362             -54      X
979872  5142362             -55      X
979873  5142362             -56      X
979874  5142362             -57      X
979875  5142362             -58      X
979876  5142362             -59      X

In [43]:
df['ID'].value_counts().sort_values(ascending=False).head(20)

5145767    61
5017242    61
5091512    61
5059942    61
5089140    61
5003783    61
5061434    61
5004106    61
5120856    61
5146385    61
5105565    61
5069425    61
5078567    61
5148524    61
5078799    61
5142738    61
5085886    61
5041979    61
5078569    61
5061019    61
Name: ID, dtype: int64

In [44]:
df.shape

(1048575, 3)

In [45]:
app.shape

(438557, 18)

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 3 columns):
 #   Column          Non-Null Count    Dtype 
---  ------          --------------    ----- 
 0   ID              1048575 non-null  int64 
 1   MONTHS_BALANCE  1048575 non-null  int64 
 2   STATUS          1048575 non-null  object
dtypes: int64(2), object(1)
memory usage: 24.0+ MB


In [47]:
df['STATUS'].value_counts()

C    442031
0    383120
X    209230
1     11090
5      1693
2       868
3       320
4       223
Name: STATUS, dtype: int64

- 0: 1-29 days past due 
- 1: 30-59 days past due 
- 2: 60-89 days overdue 
- 3: 90-119 days overdue 
- 4: 120-149 days overdue 
- 5: Overdue or bad debts, write-offs for more than 150 days 
- C: paid off that month 
- X: No loan for the month

In [48]:
1693 + 320 + 223

2236

In [49]:
# df[df.ID.]

In [82]:
oldest_credit = df[df['MONTHS_BALANCE'] == -48]

In [52]:
df.shape

(1048575, 3)

In [58]:
df.ID.nunique()

45985

In [63]:
late = df[df['STATUS'] == '1']

In [64]:
late.ID.nunique()

5201

In [65]:
app.head()

ID gender  car_owner  property_owner  children_count  annual_income  \
0  5008804      M          1               1               0         427500   
1  5008805      M          1               1               0         427500   
2  5008806      M          1               1               0         112500   
3  5008808      F          0               1               0         270000   
4  5008809      F          0               1               0         270000   

            income_type                      education         family_status  \
0               Working               Higher education        Civil marriage   
1               Working               Higher education        Civil marriage   
2               Working  Secondary / secondary special               Married   
3  Commercial associate  Secondary / secondary special  Single / not married   
4  Commercial associate  Secondary / secondary special  Single / not married   

        housing_type  days_from_birth  days_employed  mobile  work_phone  \
0   Rented apartment            12005          -4542       1           1   
1   Rented apartment            12005          -4542       1           1   
2  House / apartment            21474          -1134       1           0   
3  House / apartment            19110          -3051       1           0   
4  House / apartment            19110          -3051       1           0   

   phone  email      occupation  family_member_count  
0      0      0             NaN                    2  
1      0      0             NaN                    2  
2      0      0  Security staff                    2  
3      1      1     Sales staff                    1  
4      1      1     Sales staff                    1

In [66]:
app.education.value_counts()
#5

Secondary / secondary special    301821
Higher education                 117522
Incomplete higher                 14851
Lower secondary                    4051
Academic degree                     312
Name: education, dtype: int64

In [67]:
app.income_type.value_counts()
#5

Working                 226104
Commercial associate    100757
Pensioner                75493
State servant            36186
Student                     17
Name: income_type, dtype: int64

In [68]:
app.family_status.value_counts()
#5

Married                 299828
Single / not married     55271
Civil marriage           36532
Separated                27251
Widow                    19675
Name: family_status, dtype: int64

In [69]:
app.housing_type.value_counts()
#6

House / apartment      393831
With parents            19077
Municipal apartment     14214
Rented apartment         5974
Office apartment         3922
Co-op apartment          1539
Name: housing_type, dtype: int64

In [70]:
app.occupation.value_counts()
#18

Laborers                 78240
Core staff               43007
Sales staff              41098
Managers                 35487
Drivers                  26090
High skill tech staff    17289
Accountants              15985
Medicine staff           13520
Cooking staff             8076
Security staff            7993
Cleaning staff            5845
Private service staff     3456
Low-skill Laborers        2140
Secretaries               2044
Waiters/barmen staff      1665
Realty agents             1041
HR staff                   774
IT staff                   604
Name: occupation, dtype: int64

In [71]:
#Start from scratch with their prep code

In [72]:
apps = pd.read_csv("application_record.csv")

In [74]:
import numpy as np

In [79]:
# Convert column names to lower case
apps.columns = (apps.columns).str.lower()
# Convert days employed to years employed
apps['employed_years'] = [round(val/(-365)) if val < -365 else val/(-365)  if val < 0 else round(val/365) for val in list(apps.days_employed)]
# Convert DAYS_BIRTH to age in years
apps['age'] = (apps['days_birth']/365 * -1).apply(np.floor)
# Convert Yes and Nos to 1s and 0s whole data frame
apps.replace({'Y':1, 'N':0}, inplace=True)

In [80]:
df.head()

ID  MONTHS_BALANCE STATUS
0  5001711               0      X
1  5001711              -1      0
2  5001711              -2      0
3  5001711              -3      0
4  5001712               0      C

In [83]:
oldest_credit.groupby(['MONTHS_BALANCE', 'STATUS']).agg('count')

ID
MONTHS_BALANCE STATUS      
-48            0       3478
               1        117
               2         13
               3          7
               4          2
               5         10
               C       1063
               X       1413

- 0: 1-29 days past due
- 1: 30-59 days past due
- 2: 60-89 days overdue
- 3: 90-119 days overdue
- 4: 120-149 days overdue
- 5: Overdue or bad debts, write-offs for more than 150 days
- C: paid off that month
- X: No loan for the month

Good
Payment history of 48-61 months (6103 out of 45985)
Carry a balance and pay on time?
MEDIUM
Payment history of at least (24mo.) 20617 out of 45985. (medium)?
Possibly made payments on time, but didn’t consistently carry a balance? Paid it off early and never used again?
Poor
Roughly 20% of applicants were at least 30 days late, up to 159. If we only include write-offs it becomes statistically insignificant. (edited)

In [85]:
#If X, must be either C or 0
df['oldest_credit'] = (df['MONTHS_BALANCE'] == -48)

In [86]:
df.head()

ID  MONTHS_BALANCE STATUS  oldest_credit
0  5001711               0      X          False
1  5001711              -1      0          False
2  5001711              -2      0          False
3  5001711              -3      0          False
4  5001712               0      C          False

In [87]:
df['mid_range'] = (df['MONTHS_BALANCE'] <= -24) & (df['MONTHS_BALANCE'] > -48)

In [88]:
df.head()

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range
0  5001711               0      X          False      False
1  5001711              -1      0          False      False
2  5001711              -2      0          False      False
3  5001711              -3      0          False      False
4  5001712               0      C          False      False

In [94]:
df[(df.mid_range == True)]

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range
84       5001715             -24      X          False       True
85       5001715             -25      X          False       True
86       5001715             -26      X          False       True
87       5001715             -27      X          False       True
88       5001715             -28      X          False       True
...          ...             ...    ...            ...        ...
1048570  5150487             -25      C          False       True
1048571  5150487             -26      C          False       True
1048572  5150487             -27      C          False       True
1048573  5150487             -28      C          False       True
1048574  5150487             -29      C          False       True

[323457 rows x 5 columns]

In [93]:
#Making sure I split it up properly
df[(df.mid_range == True) & (df.oldest_credit == True)]

Empty DataFrame
Columns: [ID, MONTHS_BALANCE, STATUS, oldest_credit, mid_range]
Index: []

In [91]:
df[df.oldest_credit == True]

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range
108      5001715             -48      X           True      False
425      5001730             -48      C           True      False
661      5001739             -48      0           True      False
1500     5001786             -48      0           True      False
1837     5001798             -48      C           True      False
...          ...             ...    ...            ...        ...
1047741  5150405             -48      X           True      False
1047939  5150416             -48      0           True      False
1048178  5150459             -48      0           True      False
1048269  5150467             -48      0           True      False
1048449  5150480             -48      0           True      False

[6103 rows x 5 columns]

In [95]:
df['new_credit'] = (df['MONTHS_BALANCE'] > -24) 

In [96]:
df.head()

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit
0  5001711               0      X          False      False        True
1  5001711              -1      0          False      False        True
2  5001711              -2      0          False      False        True
3  5001711              -3      0          False      False        True
4  5001712               0      C          False      False        True

In [97]:
df[df.new_credit == True]

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit
0        5001711               0      X          False      False        True
1        5001711              -1      0          False      False        True
2        5001711              -2      0          False      False        True
3        5001711              -3      0          False      False        True
4        5001712               0      C          False      False        True
...          ...             ...    ...            ...        ...         ...
1048564  5150487             -19      C          False      False        True
1048565  5150487             -20      C          False      False        True
1048566  5150487             -21      C          False      False        True
1048567  5150487             -22      C          False      False        True
1048568  5150487             -23      C          False      False        True

[684426 rows x 6 columns]

In [98]:
df.shape

(1048575, 6)

In [99]:
df[(df.new_credit == True) & (df.mid_range == False)]

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit
0        5001711               0      X          False      False        True
1        5001711              -1      0          False      False        True
2        5001711              -2      0          False      False        True
3        5001711              -3      0          False      False        True
4        5001712               0      C          False      False        True
...          ...             ...    ...            ...        ...         ...
1048564  5150487             -19      C          False      False        True
1048565  5150487             -20      C          False      False        True
1048566  5150487             -21      C          False      False        True
1048567  5150487             -22      C          False      False        True
1048568  5150487             -23      C          False      False        True

[684426 rows x 6 columns]

In [100]:
dummies = pd.get_dummies(df['STATUS'])


In [101]:
new_df = pd.concat([df, dummies], axis=1)      


In [102]:
new_df.head()

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit  0  1  \
0  5001711               0      X          False      False        True  0  0   
1  5001711              -1      0          False      False        True  1  0   
2  5001711              -2      0          False      False        True  1  0   
3  5001711              -3      0          False      False        True  1  0   
4  5001712               0      C          False      False        True  0  0   

   2  3  4  5  C  X  
0  0  0  0  0  0  1  
1  0  0  0  0  0  0  
2  0  0  0  0  0  0  
3  0  0  0  0  0  0  
4  0  0  0  0  1  0

In [105]:
max_months = df.groupby(['ID'])['MONTHS_BALANCE'].agg('min')

In [107]:
max_months = pd.DataFrame(max_months)

In [108]:
max_months.head()

MONTHS_BALANCE
ID                     
5001711              -3
5001712             -18
5001713             -21
5001714             -14
5001715             -59

In [110]:
max_months.rename(columns={'MONTHS_BALANCE': 'id_mo_max'}, inplace=True)

In [111]:
max_months.head()

id_mo_max
ID                
5001711         -3
5001712        -18
5001713        -21
5001714        -14
5001715        -59

In [114]:
new = new_df.merge(max_months, on='ID', how='left')

In [115]:
new.head()

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit  0  1  \
0  5001711               0      X          False      False        True  0  0   
1  5001711              -1      0          False      False        True  1  0   
2  5001711              -2      0          False      False        True  1  0   
3  5001711              -3      0          False      False        True  1  0   
4  5001712               0      C          False      False        True  0  0   

   2  3  4  5  C  X  id_mo_max  
0  0  0  0  0  0  1         -3  
1  0  0  0  0  0  0         -3  
2  0  0  0  0  0  0         -3  
3  0  0  0  0  0  0         -3  
4  0  0  0  0  1  0        -18

In [117]:
new['has_had_balance'] = np.where((new_df['X'] == 1) & (new_df['C'] == 1), True, False)

In [118]:
new.head()

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit  0  1  \
0  5001711               0      X          False      False        True  0  0   
1  5001711              -1      0          False      False        True  1  0   
2  5001711              -2      0          False      False        True  1  0   
3  5001711              -3      0          False      False        True  1  0   
4  5001712               0      C          False      False        True  0  0   

   2  3  4  5  C  X  id_mo_max  has_had_balance  
0  0  0  0  0  0  1         -3            False  
1  0  0  0  0  0  0         -3            False  
2  0  0  0  0  0  0         -3            False  
3  0  0  0  0  0  0         -3            False  
4  0  0  0  0  1  0        -18            False

In [119]:
new.has_had_balance.value_counts()

False    1048575
Name: has_had_balance, dtype: int64

In [120]:
new['1'] = np.where(new['1'] == 1, 2, 0)

In [131]:
new.tail(20)

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit  \
1048555  5150487             -10      C          False      False        True   
1048556  5150487             -11      C          False      False        True   
1048557  5150487             -12      C          False      False        True   
1048558  5150487             -13      C          False      False        True   
1048559  5150487             -14      C          False      False        True   
1048560  5150487             -15      C          False      False        True   
1048561  5150487             -16      C          False      False        True   
1048562  5150487             -17      C          False      False        True   
1048563  5150487             -18      C          False      False        True   
1048564  5150487             -19      C          False      False        True   
1048565  5150487             -20      C          False      False        True   
1048566  5150487             -21      C          False      False        True   
1048567  5150487             -22      C          False      False        True   
1048568  5150487             -23      C          False      False        True   
1048569  5150487             -24      C          False       True       False   
1048570  5150487             -25      C          False       True       False   
1048571  5150487             -26      C          False       True       False   
1048572  5150487             -27      C          False       True       False   
1048573  5150487             -28      C          False       True       False   
1048574  5150487             -29      C          False       True       False   

         0  1  2  3  4  5  C  X  id_mo_max  has_had_balance  
1048555  0  0  0  0  0  0  1  0        -29            False  
1048556  0  0  0  0  0  0  1  0        -29            False  
1048557  0  0  0  0  0  0  1  0        -29            False  
1048558  0  0  0  0  0  0  1  0        -29            False  
1048559  0  0  0  0  0  0  1  0        -29            False  
1048560  0  0  0  0  0  0  1  0        -29            False  
1048561  0  0  0  0  0  0  1  0        -29            False  
1048562  0  0  0  0  0  0  1  0        -29            False  
1048563  0  0  0  0  0  0  1  0        -29            False  
1048564  0  0  0  0  0  0  1  0        -29            False  
1048565  0  0  0  0  0  0  1  0        -29            False  
1048566  0  0  0  0  0  0  1  0        -29            False  
1048567  0  0  0  0  0  0  1  0        -29            False  
1048568  0  0  0  0  0  0  1  0        -29            False  
1048569  0  0  0  0  0  0  1  0        -29            False  
1048570  0  0  0  0  0  0  1  0        -29            False  
1048571  0  0  0  0  0  0  1  0        -29            False  
1048572  0  0  0  0  0  0  1  0        -29            False  
1048573  0  0  0  0  0  0  1  0        -29            False  
1048574  0  0  0  0  0  0  1  0        -29            False

In [122]:
new['1'].value_counts()

0    1037485
2      11090
Name: 1, dtype: int64

In [123]:
new['2'] = np.where(new['2'] == 1, 3, 0)

In [124]:
new['3'] = np.where(new['3'] == 1, 4, 0)

In [125]:
new['4'] = np.where(new['4'] == 1, 5, 0)

In [126]:
new['5'] = np.where(new['5'] == 1, 6, 0)

In [127]:
new['5'].value_counts()

0    1046882
6       1693
Name: 5, dtype: int64

In [129]:
new[new.ID == 5001711]

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit  0  1  \
0  5001711               0      X          False      False        True  0  0   
1  5001711              -1      0          False      False        True  1  0   
2  5001711              -2      0          False      False        True  1  0   
3  5001711              -3      0          False      False        True  1  0   

   2  3  4  5  C  X  id_mo_max  has_had_balance  
0  0  0  0  0  0  1         -3            False  
1  0  0  0  0  0  0         -3            False  
2  0  0  0  0  0  0         -3            False  
3  0  0  0  0  0  0         -3            False

In [133]:
new['score'] = new['0'] + new['1'] + new['2'] + new['3'] + new['4'] + new['5']

In [134]:
new.head()

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit  0  1  \
0  5001711               0      X          False      False        True  0  0   
1  5001711              -1      0          False      False        True  1  0   
2  5001711              -2      0          False      False        True  1  0   
3  5001711              -3      0          False      False        True  1  0   
4  5001712               0      C          False      False        True  0  0   

   2  3  4  5  C  X  id_mo_max  has_had_balance  score  
0  0  0  0  0  0  1         -3            False      0  
1  0  0  0  0  0  0         -3            False      1  
2  0  0  0  0  0  0         -3            False      1  
3  0  0  0  0  0  0         -3            False      1  
4  0  0  0  0  1  0        -18            False      0

In [135]:
new.score.value_counts()

0    651261
1    383120
2     11090
6      1693
3       868
4       320
5       223
Name: score, dtype: int64

In [138]:
agg_score = new.groupby(['ID']).score.sum()

In [139]:
total = pd.DataFrame(agg_score)

In [150]:
total.head()

score
ID            
5001711      3
5001712     10
5001713      0
5001714      0
5001715      0

In [148]:
total.score.value_counts()

0      5953
6      3778
4      2935
5      2820
3      2691
       ... 
296       1
168       1
136       1
104       1
109       1
Name: score, Length: 118, dtype: int64

In [151]:
max(total.score)

296

In [141]:
new = new.merge(total, on='ID', how='left')

In [142]:
new.head()

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit  0  1  \
0  5001711               0      X          False      False        True  0  0   
1  5001711              -1      0          False      False        True  1  0   
2  5001711              -2      0          False      False        True  1  0   
3  5001711              -3      0          False      False        True  1  0   
4  5001712               0      C          False      False        True  0  0   

   2  3  4  5  C  X  id_mo_max  has_had_balance  score_x  score_y  
0  0  0  0  0  0  1         -3            False        0        3  
1  0  0  0  0  0  0         -3            False        1        3  
2  0  0  0  0  0  0         -3            False        1        3  
3  0  0  0  0  0  0         -3            False        1        3  
4  0  0  0  0  1  0        -18            False        0       10

In [152]:
new['score_y'].value_counts()

0      135552
6       79898
10      60013
12      58098
5       53257
        ...  
149        31
116        31
68         26
89         21
109        19
Name: score_y, Length: 118, dtype: int64

In [155]:
scores_per_id = new.groupby(['ID']).score_y.agg('max')

In [156]:
scores = pd.DataFrame(scores_per_id)

In [157]:
scores.head()

score_y
ID              
5001711        3
5001712       10
5001713        0
5001714        0
5001715        0

In [158]:
scores.rename(columns={'score_y': 'total'}, inplace=True)

In [159]:
scores.total > 100

ID
5001711    False
5001712    False
5001713    False
5001714    False
5001715    False
           ...  
5150482    False
5150483    False
5150484    False
5150485    False
5150487    False
Name: total, Length: 45985, dtype: bool

In [161]:
scores.total <= 100 

ID
5001711    True
5001712    True
5001713    True
5001714    True
5001715    True
           ... 
5150482    True
5150483    True
5150484    True
5150485    True
5150487    True
Name: total, Length: 45985, dtype: bool

In [162]:
scores.total.value_counts()

0      5953
6      3778
4      2935
5      2820
3      2691
       ... 
296       1
168       1
136       1
104       1
109       1
Name: total, Length: 118, dtype: int64

In [163]:
scores.shape

(45985, 1)

In [164]:
scores.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45985 entries, 5001711 to 5150487
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   total   45985 non-null  int64
dtypes: int64(1)
memory usage: 718.5 KB


In [168]:
scores[scores.total < 10]

total
ID            
5001711      3
5001713      0
5001714      0
5001715      0
5001719      2
...        ...
5150480      6
5150481      0
5150483      0
5150485      2
5150487      0

[29364 rows x 1 columns]

In [169]:
scores[scores.total >= 10]

total
ID            
5001712     10
5001717     17
5001718     28
5001720     43
5001737     12
...        ...
5150473     35
5150476     17
5150477     17
5150482     12
5150484     12

[16621 rows x 1 columns]

In [170]:
scores[scores.total < 5]

total
ID            
5001711      3
5001713      0
5001714      0
5001715      0
5001719      2
...        ...
5150479      0
5150481      0
5150483      0
5150485      2
5150487      0

[16290 rows x 1 columns]

In [171]:
scores[(scores.total >= 5) & (scores.total <= 10)]

total
ID            
5001712     10
5001723      8
5001725      7
5001726      6
5001729      6
...        ...
5150428      7
5150429      7
5150467      6
5150468      6
5150480      6

[15441 rows x 1 columns]

In [172]:
age = new[['ID', 'oldest_credit', 'mid_range', 'new_credit']]

In [173]:
age.head()

ID  oldest_credit  mid_range  new_credit
0  5001711          False      False        True
1  5001711          False      False        True
2  5001711          False      False        True
3  5001711          False      False        True
4  5001712          False      False        True

In [174]:
df2 = age.merge(scores, on='ID', how='left')

In [175]:
df2.head()

ID  oldest_credit  mid_range  new_credit  total
0  5001711          False      False        True      3
1  5001711          False      False        True      3
2  5001711          False      False        True      3
3  5001711          False      False        True      3
4  5001712          False      False        True     10

In [ ]:
#For MVP

In [178]:
new.head()

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit  0  1  \
0  5001711               0      X          False      False        True  0  0   
1  5001711              -1      0          False      False        True  1  0   
2  5001711              -2      0          False      False        True  1  0   
3  5001711              -3      0          False      False        True  1  0   
4  5001712               0      C          False      False        True  0  0   

   2  3  4  5  C  X  id_mo_max  has_had_balance  score_x  score_y  
0  0  0  0  0  0  1         -3            False        0        3  
1  0  0  0  0  0  0         -3            False        1        3  
2  0  0  0  0  0  0         -3            False        1        3  
3  0  0  0  0  0  0         -3            False        1        3  
4  0  0  0  0  1  0        -18            False        0       10

In [180]:
cred_hist = new.groupby(['ID']).MONTHS_BALANCE.agg('min')

In [181]:
cred_hist = pd.DataFrame(cred_hist)

In [182]:
cred_hist.head()

MONTHS_BALANCE
ID                     
5001711              -3
5001712             -18
5001713             -21
5001714             -14
5001715             -59

In [194]:
cred_hist['time_as_cust_score'] = np.where(cred_hist['MONTHS_BALANCE'] > -24, 0, np.where(cred_hist['MONTHS_BALANCE'] > -47, -10, -20))

In [195]:
cred_hist.head()

MONTHS_BALANCE  time_as_cust_score
ID                                         
5001711              -3                   0
5001712             -18                   0
5001713             -21                   0
5001714             -14                   0
5001715             -59                 -20

In [185]:
new.head()

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit  0  1  \
0  5001711               0      X          False      False        True  0  0   
1  5001711              -1      0          False      False        True  1  0   
2  5001711              -2      0          False      False        True  1  0   
3  5001711              -3      0          False      False        True  1  0   
4  5001712               0      C          False      False        True  0  0   

   2  3  4  5  C  X  id_mo_max  has_had_balance  score_x  score_y  
0  0  0  0  0  0  1         -3            False        0        3  
1  0  0  0  0  0  0         -3            False        1        3  
2  0  0  0  0  0  0         -3            False        1        3  
3  0  0  0  0  0  0         -3            False        1        3  
4  0  0  0  0  1  0        -18            False        0       10

In [186]:
new[new['MONTHS_BALANCE'] <= -48]

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit  \
108      5001715             -48      X           True      False       False   
109      5001715             -49      X          False      False       False   
110      5001715             -50      X          False      False       False   
111      5001715             -51      X          False      False       False   
112      5001715             -52      X          False      False       False   
...          ...             ...    ...            ...        ...         ...   
1048271  5150467             -50      0          False      False       False   
1048272  5150467             -51      0          False      False       False   
1048273  5150467             -52      0          False      False       False   
1048449  5150480             -48      0           True      False       False   
1048450  5150480             -49      0          False      False       False   

         0  1  2  3  4  5  C  X  id_mo_max  has_had_balance  score_x  score_y  
108      0  0  0  0  0  0  0  1        -59            False        0        0  
109      0  0  0  0  0  0  0  1        -59            False        0        0  
110      0  0  0  0  0  0  0  1        -59            False        0        0  
111      0  0  0  0  0  0  0  1        -59            False        0        0  
112      0  0  0  0  0  0  0  1        -59            False        0        0  
...     .. .. .. .. .. .. .. ..        ...              ...      ...      ...  
1048271  1  0  0  0  0  0  0  0        -52            False        1        6  
1048272  1  0  0  0  0  0  0  0        -52            False        1        6  
1048273  1  0  0  0  0  0  0  0        -52            False        1        6  
1048449  1  0  0  0  0  0  0  0        -49            False        1        6  
1048450  1  0  0  0  0  0  0  0        -49            False        1        6  

[40692 rows x 18 columns]

In [187]:
new.shape

(1048575, 18)

In [189]:
new[(new['MONTHS_BALANCE'] <= -24) & (new['MONTHS_BALANCE'] > -48)]

ID  MONTHS_BALANCE STATUS  oldest_credit  mid_range  new_credit  \
84       5001715             -24      X          False       True       False   
85       5001715             -25      X          False       True       False   
86       5001715             -26      X          False       True       False   
87       5001715             -27      X          False       True       False   
88       5001715             -28      X          False       True       False   
...          ...             ...    ...            ...        ...         ...   
1048570  5150487             -25      C          False       True       False   
1048571  5150487             -26      C          False       True       False   
1048572  5150487             -27      C          False       True       False   
1048573  5150487             -28      C          False       True       False   
1048574  5150487             -29      C          False       True       False   

         0  1  2  3  4  5  C  X  id_mo_max  has_had_balance  score_x  score_y  
84       0  0  0  0  0  0  0  1        -59            False        0        0  
85       0  0  0  0  0  0  0  1        -59            False        0        0  
86       0  0  0  0  0  0  0  1        -59            False        0        0  
87       0  0  0  0  0  0  0  1        -59            False        0        0  
88       0  0  0  0  0  0  0  1        -59            False        0        0  
...     .. .. .. .. .. .. .. ..        ...              ...      ...      ...  
1048570  0  0  0  0  0  0  1  0        -29            False        0        0  
1048571  0  0  0  0  0  0  1  0        -29            False        0        0  
1048572  0  0  0  0  0  0  1  0        -29            False        0        0  
1048573  0  0  0  0  0  0  1  0        -29            False        0        0  
1048574  0  0  0  0  0  0  1  0        -29            False        0        0  

[323457 rows x 18 columns]

In [196]:
cred_hist.time_as_cust_score.value_counts()

 0     21979
-10    17005
-20     7001
Name: time_as_cust_score, dtype: int64

In [ ]:
#Code block to share with Anthony for points system based on length of credit history.
# cred_hist = new.groupby(['ID']).MONTHS_BALANCE.agg('min')
# cred_hist = pd.DataFrame(cred_hist)
# cred_hist['time_as_cust_score'] = np.where(cred_hist['MONTHS_BALANCE'] > -24, 0, np.where(cred_hist['MONTHS_BALANCE'] > -47, -10, -20))